### PR TITLE
2FA Base branch

### PR DESCRIPTION
**Pull request by berkeleycole**
# Summary

Enable two factor authentication via email. The user would provide their email/username and password (existing functionality) then be prompted to input a 6 digit code delivered to their email (new functionality) in order to login. The 6 digit code will be generated as an TOTP (Time-based One Time Password) code.

### Benefit

Many companies are now requiring 2FA on all accounts and platforms their employees may use. The addition of this feature would allow companies using oxidauth to comply with this requirement.

### Technical Flow
Each user is assigned a static auth key at creation (stored in new table totp_secrets). From the frontend login screen, a user provides username & password combo, if password is correct:

- a permission limited jwt is sent to the browser, allowing the user to access the 2FA code input screen
- a 6 digit code is created from the secret key + current time, then emailed to the user by oxidauth
- user inputs the 6 digit code from their email which is diffed against the secrety key + current time
- if 5 minutes have passed, the new code will not match the user entered code, and login will fail. Returns an error.
- if user input code does match the new code for any reason, login will fail. Returns an error
- if user requests a new code, the 6 digit code is recreated with new current time, and emailed to the user
- user can continue to try codes until the temporary jwt times out (or, max number of tries is reached, if we implement)
- if user input matches new code, a jwt is returned and user is logged in

### New Endpoints
- [POST] /totp/validate - body supplies a code for verification and a user id
- [GET] /totp - requests new code & email

### New Frontend Pages
- Email code entry form

### Other changes
- OxidAuth Authority will now have an added property of `require-2fa`, which, if true, will create a 2fa code for each new user and impact the sign in flow.
- The authenticate logic will now have two paths. If the username password authority does not require 2FA, the flow remains unchanged and upon username password validation, returns the full JWT. We are adding a branch in this logic to detect if 2FA is required, and instead trigger the email send and return a limited JWT with only the permissions to see the email code page.

### Libraries
Code generation library: [Boring Auth](https://docs.rs/boringauth) This package is a library designed to provide out-of-box HOTP and TOTP clients to generate one-time passwords.

### Database Migrations
Table: totp_secrets
Columns: id, user_id, totp_secret, created_at, updated_at

### Research & Alternatives
- Alternate libraries: rust-otp, otps, libreauth
- HOTP spec hotp RFC
- a company that did this (gosquared)[https://www.gosquared.com/blog/building-two-factor-authentication]
- a resource - used multiple articles here (onelogin)[https://www.onelogin.com/learn/otp-totp-hotp]

### UI
Screenshot 2024-04-12-login

Screenshot 2024-04-12 at 10 46 52 AM

### CODE REVIEW

File: oxidauth-http/src/main.rs [hunk 1]
```diff
@@ -32,14 +29,10 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
         .call(&BootstrapParams)
         .await?;
 
-    info!("starting server...");
```
Comment (berkeleycole) [inline oxidauth-http/src/main.rs:35]: Keep these

File: oxidauth-http/src/provider/mod.rs [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
 use oxidauth_kernel::error::BoxedError;
+use oxidauth_kernel::mailer::service::IntoSenderService;
 pub use oxidauth_kernel::provider::Provider;
 use oxidauth_postgres::Database;
+use oxidauth_usecases::mailer::smtp::Smtp;
 
 pub async fn setup() -> Result<Provider, BoxedError> {
     let mut provider = Provider::new();
 
+    // DB setup ---------------------------------
```
Comment (berkeleycole) [inline oxidauth-http/src/provider/mod.rs:16]: maybe turn into a scope

File: oxidauth-usecases/src/totp/validate.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        // prepare TOTP secret params
+        let secret_params = SelectTOTPSecretByUserId { user_id };
+
+        // get the secret key for the user by id
+        let secret_by_user_id: TOTPSecret = self
+            .secret
+            .call(&secret_params)
+            .await?;
+
+        let valid = boringauth::oath::TOTPBuilder::new()
```
Comment (berkeleycole) [inline oxidauth-usecases/src/totp/validate.rs:100]: don't import like this

File: oxidauth-usecases/src/totp/validate.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            .ascii_key(&secret_by_user_id.secret)
+            .period(300)
+            .finalize()
+            .unwrap()
+            .is_valid(&req.code);
+
+        if !valid {
+            return Err("Code was not valid".into());
```
Comment (berkeleycole) [inline oxidauth-usecases/src/totp/validate.rs:108]:
```suggestion
            return Err("invalid totp code".into());
```

File: oxidauth-usecases/src/totp/validate.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        // BUILD JWT ----------------------------------------
+
+        let authority_params = FindAuthorityByClientKey {
+            client_key: req.client_key,
+        };
+
+        let authority = self
+            .authority_by_client_key
+            .call(&authority_params)
```
Comment (berkeleycole) [inline oxidauth-usecases/src/totp/validate.rs:119]: inline this struct

File: oxidauth-usecases/src/totp/validate.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            .with_subject(user_id)
+            .with_issuer("oxidauth".to_owned())
+            .with_not_before_from(Duration::from_secs(0));
+
+        let refresh_token_exp_at = epoch_from_now(
+            authority
+                .settings
+                .refresh_token_ttl,
+        )
+        .map_err(|err| {
```
Comment (berkeleycole) [inline oxidauth-usecases/src/totp/validate.rs:152]: tab this in

File: oxidauth-kernel/src/totp/generate.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        Error = BoxedError,
+    >,
+>;
+
+pub trait GenerateTOTPTrait:
+    for<'a> Service<
+    &'a GenerateTOTP,
+    Response = TOTPGenerationRes,
+    Error = BoxedError,
+>
```
Comment (berkeleycole) [inline oxidauth-kernel/src/totp/generate.rs:22]:
```suggestion
        &'a GenerateTOTP,
        Response = TOTPGenerationRes,
        Error = BoxedError,
    >
```

File: oxidauth-kernel/src/totp_secrets/create_totp_secret.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+use crate::dev_prelude::*;
+use uuid::Uuid;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CreateTotpSecretResponse {
+    pub success: bool,
+}
+
+pub struct InsertTotpSecretParams {
```
Comment (berkeleycole) [inline oxidauth-kernel/src/totp_secrets/create_totp_secret.rs:9]: move to repository

File: oxidauth-kernel/src/totp_secrets/mod.rs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+pub mod create_totp_secret;
+pub mod select_totp_secret_by_user_id;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TOTPSecret {
+    pub secret: String,
+}
+
+impl fmt::Display for TOTPSecret {
```
Comment (berkeleycole) [inline oxidauth-kernel/src/totp_secrets/mod.rs:13]: remove if not needed, or display secret

File: oxidauth-kernel/src/totp_secrets/select_totp_secret_by_user_id.rs [hunk 10]
```diff
@@ -0,0 +1,37 @@
+use std::sync::Arc;
```
Comment (berkeleycole) [inline oxidauth-kernel/src/totp_secrets/select_totp_secret_by_user_id.rs:1]: rename this file and trait to find

File: oxidauth-kernel/src/totp_secrets/select_totp_secret_by_user_id.rs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+#[derive(Debug)]
+pub struct SelectTOTPSecretByUserIdError;
+
+impl fmt::Display for SelectTOTPSecretByUserIdError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "could not find TOTP secret for user id"
```
Comment (berkeleycole) [inline oxidauth-kernel/src/totp_secrets/select_totp_secret_by_user_id.rs:32]: maybe add user id to this error message if possible

File: oxidauth-postgres/Cargo.toml [hunk 12]
```diff
@@ -26,3 +26,4 @@ tracing-log = "0.2.0"
 tracing-subscriber = { version = "0.3.18", features = ["registry", "env-filter"] }
 uuid = { version = "1.4.0", features = ["v4", "serde"] }
 async-recursion = "1.0.5"
+rand = "0.8.5"
```
Comment (berkeleycole) [inline oxidauth-postgres/Cargo.toml:29]:
```suggestion

```

File: oxidauth-postgres/src/totp_secrets/insert_totp_secret/mod.rs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        &self,
+        params: &'a InsertTotpSecretParams,
+    ) -> Result<Self::Response, Self::Error> {
+        let result = sqlx::query_as::<_, PgTotpSecret>(include_str!(
+            "./insert_totp_secret.sql"
+        ))
+        .bind(&params.user_id)
+        .bind(&params.secret_key)
+        .fetch_one(&self.pool)
+        .await?;
```
Comment (berkeleycole) [inline oxidauth-postgres/src/totp_secrets/insert_totp_secret/mod.rs:29]:
```suggestion
            .bind(&params.user_id)
            .bind(&params.secret_key)
            .fetch_one(&self.pool)
            .await?;
```

File: oxidauth-postgres/src/totp_secrets/insert_totp_secret/mod.rs [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+        let _: TOTPSecret = result.try_into()?;
```
Comment (berkeleycole) [inline oxidauth-postgres/src/totp_secrets/insert_totp_secret/mod.rs:31]: circle back on this

File: oxidauth-postgres/src/totp_secrets/select_totp_secret_by_user_id/mod.rs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+pub async fn select_totp_secret_by_user_id_query(
+    conn: &mut PgConnection,
+    user_id: Uuid,
+) -> Result<TOTPSecret, BoxedError> {
+    let result = sqlx::query_as::<_, PgTotpSecret>(include_str!(
+        "./select_totp_secret_by_user_id.sql"
+    ))
+    .bind(user_id)
+    .fetch_one(conn)
+    .await?;
```
Comment (berkeleycole) [inline oxidauth-postgres/src/totp_secrets/select_totp_secret_by_user_id/mod.rs:42]:
```suggestion
        .bind(user_id)
        .fetch_one(conn)
        .await?;
```

File: oxidauth-postgres/src/totp_secrets/select_totp_secret_by_user_id/select_totp_secret_by_user_id.sql [hunk 16]
```diff
@@ -0,0 +1,3 @@
+SELECT totp_secret
```
Comment (berkeleycole) [inline oxidauth-postgres/src/totp_secrets/select_totp_secret_by_user_id/select_totp_secret_by_user_id.sql:1]: Change to select everything

File: oxidauth-usecases/src/auth/authenticate.rs [hunk 17]
```diff
@@ -115,41 +128,55 @@ where
...
-            .build()
-            .map_err(|err| {
-                format!(
-                    "unable to build jwt: {:?}",
-                    err
+            .with_not_before_from(Duration::from_secs(0));
+
+        // CHECK FOR 2FA REQUIREMENT AND START 2FA FLOW ---------
+        if authority.settings.require_2fa {
+            info!("Login requires 2FA");
```
Comment (berkeleycole) [inline oxidauth-usecases/src/auth/authenticate.rs:145]:
```suggestion
            info!("login requires 2FA");
```

File: oxidauth-usecases/src/bootstrap/mod.rs [hunk 18]
```diff
@@ -238,12 +240,35 @@ async fn first_or_create_public_key(
...
 
 pub const ADMIN_PERMISSION: &str = "oxidauth:**:**";
+pub const TOTP_PERMISSION: &str = "oxidauth:totp_code:validate";
 
 #[tracing::instrument(skip_all)]
 async fn first_or_create_permissions(
     permission_by_name: &FindPermissionByPartsService,
     create_permission: &CreatePermissionService,
 ) -> Result<Permission, BoxedError> {
+    let permission_name = TOTP_PERMISSION.to_owned();
```
Comment (berkeleycole) [inline oxidauth-usecases/src/bootstrap/mod.rs:250]:
```suggestion
    // TODO(dewey4iv): currently not returning but might want to later
    
    let permission_name = TOTP_PERMISSION.to_owned();
```

File: oxidauth-usecases/src/bootstrap/mod.rs [hunk 19]
```diff
@@ -329,6 +354,10 @@ async fn add_admin_permission_to_admin_role(
 
 pub const DEFAULT_JWT_TTL: Duration = Duration::from_secs(60 * 2);
 
+pub const DEFAULT_2FA: bool = false;
```
Comment (berkeleycole) [inline oxidauth-usecases/src/bootstrap/mod.rs:357]: remove extra line

File: oxidauth-usecases/src/bootstrap/mod.rs [hunk 20]
```diff
@@ -362,6 +391,8 @@ async fn first_or_create_authority(
                     let authority_settings = AuthoritySettings {
                         jwt_ttl: DEFAULT_JWT_TTL,
                         refresh_token_ttl: DEFAULT_REFRESH_TOKEN_TTL,
+                        require_2fa: DEFAULT_2FA,
+                        totp_jwt_ttl: DEFAULT_TOTP_TOKEN_TTL,
```
Comment (berkeleycole) [inline oxidauth-usecases/src/bootstrap/mod.rs:395]: This should be an Option of totp_jwt_ttl - edit code to work with that and set default to none

File: changelogs/187442172-totp-secrets-migration-and-insert.md [hunk 21]
```diff
@@ -0,0 +1,4 @@
+[187442172](https://www.pivotaltracker.com/story/show/187442172)- TOTP secrets
+    - totp secrets table migration
+    - totp secrets generate random key
+    - totp secrets insert into DB on user creation
```
Comment (berkeleycole) [inline changelogs/187442172-totp-secrets-migration-and-insert.md:4]:
```suggestion
    - totp secrets insert into DB on user creation
    
```

File: changelogs/187452278-2fa-authority-setting.md [hunk 22]
```diff
@@ -0,0 +1 @@
+- [187452278](https://www.pivotaltracker.com/story/show/187452278) - adds 2FA setting to authority
```
Comment (berkeleycole) [inline changelogs/187452278-2fa-authority-setting.md:1]:
```suggestion
- [187452278](https://www.pivotaltracker.com/story/show/187452278) - adds 2FA setting to authority

```

File: changelogs/187572686-authorize-with-2fa-logic.md [hunk 23]
```diff
@@ -0,0 +1,3 @@
+- [187572686](https://www.pivotaltracker.com/story/show/187572686) - 2FA in authorize flow
+    - on username password validation, checks for whether or not 2FA is required
+    - returns JWT with limited permission
```
Comment (berkeleycole) [inline changelogs/187572686-authorize-with-2fa-logic.md:3]:
```suggestion
    - returns JWT with limited permission

```

File: example.env [hunk 24]
```diff
@@ -3,4 +3,4 @@ RUST_DEV_IMAGE_VERSION="v1.76.0"
 POSTGRES_DATABASE=oxidauth
 POSTGRES_USER=oxidauth
 POSTGRES_PASSWORD=oxidauth
-OXIDAUTH_USERNAME_PASSWORD_PPPER=pepper
+OXIDAUTH_USERNAME_PASSWORD_PPPER=pepper
```
Comment (berkeleycole) [inline example.env:6]:
```suggestion
OXIDAUTH_USERNAME_PASSWORD_PEPPER=pepper
```

File: oxidauth-postgres/migrations/20240619063635_modify_user_authority_user_identifier_unique.sql [hunk 25]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE user_authorities DROP CONSTRAINT user_authorities_user_identifier_key;
```
Comment (berkeleycole) [inline oxidauth-postgres/migrations/20240619063635_modify_user_authority_user_identifier_unique.sql:1]:
```suggestion
ALTER TABLE user_authorities
DROP CONSTRAINT user_authorities_user_identifier_key;
```

File: oxidauth-usecases/src/auth/authenticate.rs [hunk 26]
```diff
@@ -115,41 +131,126 @@ where
...
+                    .user_by_id
+                    .call(&FindUserById {
+                        user_id: user_authority.user_id,
+                    })
+                    .await?;
+
+                info!("login requires 2FA");
+
+                // Return permission for viewing just the email code form in frontend
+                const TOTP_PERMISSION: &str = "oxidauth:totp_code:validate";
```
Comment (berkeleycole) [inline oxidauth-usecases/src/auth/authenticate.rs:163]: It looks like this is defined in multiple places...

File: oxidauth-usecases/src/auth/authenticate.rs [hunk 27]
```diff
@@ -115,41 +131,126 @@ where
...
+                        user_id: user_authority.user_id,
+                    })
+                    .await?;
+
+                let now = match SystemTime::now().duration_since(UNIX_EPOCH) {
+                    Ok(n) => n,
+                    Err(_) => {
+                        return Err("Time is before 1970".into());
+                    },
+                };
```
Comment (berkeleycole) [inline oxidauth-usecases/src/auth/authenticate.rs:184]: Use map_err

File: changelogs/187442172-totp-secrets-migration-and-insert.md (hunk 21 above)
Comment (berkeleycole) [inline changelogs/187442172-totp-secrets-migration-and-insert.md:4]:
```suggestion
    - totp secrets insert into DB on user creation

```

File: changelogs/187572686-authorize-with-2fa-logic.md (hunk 23 above)
Comment (berkeleycole) [inline changelogs/187572686-authorize-with-2fa-logic.md:3]:
```suggestion
    - returns JWT with limited permission

```

**Review by dewey4iv — APPROVED** [review]

**Pull request merged**